### ISSUE TITLE
Tomcat Probe issues - Tomcat 9.0.85 with JDK 17

### ISSUE THREAD

**Issue by khravi1979** [psi-probe/psi-probe#3153]
Hi All,

We are using Tomcat 9.0.85 version with JDK 17 (17.0.10). Tomcat is running fine while launching Probe, we are getting this error,

"Application encountered an unexpected error. Please check the Tomcat logs. We will greatly appreciate it if you share the information with us. Sharing this information through our [issue tracker](https://github.com/psi-probe/psi-probe/issues) or our [discussion board](http://groups.google.com/group/psi-probe-discuss) will help us find and correct this problem as soon as possible!"

In the logs, we see this error,

04-Mar-2024 14:20:37.721 SEVERE [main] org.apache.catalina.core.ApplicationContext.log Servlet.init() for servlet [probe] threw exception
        org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'handlerMapping' defined in ServletContext resource [/WEB-INF/spring-probe-servlet.xml]: Initi
alization of bean failed; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name '/appruntimeinfo.ajax' defined in ServletContext resourc
e [/WEB-INF/spring-probe-controllers.xml]: Cannot resolve reference to bean 'statsCollection' while setting bean property 'statsCollection'; nested exception is org.springframework.beans.fa
ctory.BeanCreationException: Error creating bean with name 'statsCollection' defined in ServletContext resource [/WEB-INF/spring-probe-stats.xml]: Invocation of init method failed; nested e
xception is java.lang.ExceptionInInitializerError

Kindly help us in fixing this issue.

**Comment by hazendaz** [comment]
That area might be related to your Catalina setup.  Please make sure your configs are up to date with tomcats.  Look at our Readme as we added some info tomcat has and requires to work.  While your stack doesn't state same. It feels like it's a possibility of not having all tomcat config in some way.

If still an issue I can try same version tonight.  I've been testing on older for a while (9.0.52).

Also are you on the latest psi probe?  4.1.4?

________________________________
From: khravi1979 ***@***.***>
Sent: Monday, March 4, 2024 9:33:50 AM
To: psi-probe/psi-probe ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [psi-probe/psi-probe] Tomcat Probe issues - Tomcat 9.0.85 with JDK 17 (Issue #3153)


Hi All,

We are using Tomcat 9.0.85 version with JDK 17 (17.0.10). Tomcat is running fine while launching Probe, we are getting this error,

"Application encountered an unexpected error. Please check the Tomcat logs. We will greatly appreciate it if you share the information with us. Sharing this information through our issue tracker<https://github.com/psi-probe/psi-probe/issues> or our discussion board<http://groups.google.com/group/psi-probe-discuss> will help us find and correct this problem as soon as possible!"

In the logs, we see this error,

04-Mar-2024 14:20:37.721 SEVERE [main] org.apache.catalina.core.ApplicationContext.log Servlet.init() for servlet [probe] threw exception
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'handlerMapping' defined in ServletContext resource [/WEB-INF/spring-probe-servlet.xml]: Initi
alization of bean failed; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name '/appruntimeinfo.ajax' defined in ServletContext resourc
e [/WEB-INF/spring-probe-controllers.xml]: Cannot resolve reference to bean 'statsCollection' while setting bean property 'statsCollection'; nested exception is org.springframework.beans.fa
ctory.BeanCreationException: Error creating bean with name 'statsCollection' defined in ServletContext resource [/WEB-INF/spring-probe-stats.xml]: Invocation of init method failed; nested e
xception is java.lang.ExceptionInInitializerError

Kindly help us in fixing this issue.

**Comment by khravi1979** [comment]
Hi Jeremy,

Yes, we are using latest version of Probe.

Regards,
Ravikumar Haritharan
Mobile :- +91 9444126426
EMail:- ***@***.******@***.***>
***@***.***


From: Jeremy Landis ***@***.***>
Sent: Monday, March 4, 2024 9:15 PM
To: psi-probe/psi-probe ***@***.***>
Cc: RAVIKUMAR KUTTY HARITHARAN ***@***.***>; Author ***@***.***>
Subject: [EXTERNAL] Re: [psi-probe/psi-probe] Tomcat Probe issues - Tomcat 9.0.85 with JDK 17 (Issue #3153)

That area might be related to your Catalina setup. Please make sure your configs are up to date with tomcats. Look at our Readme as we added some info tomcat has and requires to work. While your stack doesn't state same. It feels like it's
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender
This message came from outside your organization.
    Report Suspicious  <https://us-phishalarm-ewt.proofpoint.com/EWT/v1/Dq81_Y1A2Q!eNONmMCJTy6EFq_I60oZy3JxyyB-oYKQU6ZfkZOUiFqX9LUv0BDX4y9SrVvLxcjK3ldgGr5FtQ9jwfvk1jsr5kis4gRcy3Su7fT_Uq-9kQo$>   ‌
ZjQcmQRYFpfptBannerEnd
That area might be related to your Catalina setup. Please make sure your configs are up to date with tomcats. Look at our Readme as we added some info tomcat has and requires to work. While your stack doesn't state same. It feels like it's a possibility of not having all tomcat config in some way.

If still an issue I can try same version tonight. I've been testing on older for a while (9.0.52).

Also are you on the latest psi probe? 4.1.4?

________________________________
From: khravi1979 ***@***.***>
Sent: Monday, March 4, 2024 9:33:50 AM
To: psi-probe/psi-probe ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [psi-probe/psi-probe] Tomcat Probe issues - Tomcat 9.0.85 with JDK 17 (Issue #3153)


Hi All,

We are using Tomcat 9.0.85 version with JDK 17 (17.0.10). Tomcat is running fine while launching Probe, we are getting this error,

"Application encountered an unexpected error. Please check the Tomcat logs. We will greatly appreciate it if you share the information with us. Sharing this information through our issue tracker<https://github.com/psi-probe/psi-probe/issues><https://github.com/psi-probe/psi-probe/issues%3E> or our discussion board<http://groups.google.com/group/psi-probe-discuss><http://groups.google.com/group/psi-probe-discuss%3E> will help us find and correct this problem as soon as possible!"

In the logs, we see this error,

04-Mar-2024 14:20:37.721 SEVERE [main] org.apache.catalina.core.ApplicationContext.log Servlet.init() for servlet [probe] threw exception
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'handlerMapping' defined in ServletContext resource [/WEB-INF/spring-probe-servlet.xml]: Initi
alization of bean failed; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name '/appruntimeinfo.ajax' defined in ServletContext resourc
e [/WEB-INF/spring-probe-controllers.xml]: Cannot resolve reference to bean 'statsCollection' while setting bean property 'statsCollection'; nested exception is org.springframework.beans.fa
ctory.BeanCreationException: Error creating bean with name 'statsCollection' defined in ServletContext resource [/WEB-INF/spring-probe-stats.xml]: Invocation of init method failed; nested e
xception is java.lang.ExceptionInInitializerError

Kindly help us in fixing this issue.

**Comment by khravi1979** [comment]
we are seeing a strange issue with probe on Tomcat 9.0.85. after restart of tomcat, probe URL gives this below attached error (image attached)
<img width="940" alt="probe" src="https://github.com/psi-probe/psi-probe/assets/162144474/05ca3b4a-bcce-433c-8591-476773905fd3">

But if we remove probe folder in webbapps and refresh probe URL, the URL is coming OK.

Kindly let us know how to fix this issue.

**Comment by hazendaz** [comment]
What do your logs show?

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by khravi1979** [comment]
04-Mar-2024 14:20:37.721 SEVERE [main] org.apache.catalina.core.ApplicationContext.log Servlet.init() for servlet [probe] threw exception
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'handlerMapping' defined in ServletContext resource [/WEB-INF/spring-probe-servlet.xml]: Initi
alization of bean failed; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name '/appruntimeinfo.ajax' defined in ServletContext resourc
e [/WEB-INF/spring-probe-controllers.xml]: Cannot resolve reference to bean 'statsCollection' while setting bean property 'statsCollection'; nested exception is org.springframework.beans.fa
ctory.BeanCreationException: Error creating bean with name 'statsCollection' defined in ServletContext resource [/WEB-INF/spring-probe-stats.xml]: Invocation of init method failed; nested e
xception is java.lang.ExceptionInInitializerError

**Comment by hazendaz** [comment]
Can you confirm your tomcat configuration files are up to date?  I think this is related to the following: NOTE: Picked up JDK_JAVA_OPTIONS:  --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED --add-opens=java.rmi/sun.rmi.transport=ALL-UNNAMED
Tomcat exports all that in their configs.  Its needed for the stats stuff to work.

Now I am using a older 9.0.52 but I don't think that ever changed in what they export.  So I think that should be looked at.  I'll try upgrading this weekend to 9.0.87 for my testing and see if I see any such issues.  I've also got a few fixes but not related to that which will get released this weekend for 4.1.5.

**Comment by khravi1979** [comment]
We added this in startup script of Tomcat and restarted Tomcat,, the result is the same
we need to remove probe folder inside webapps and then the URL loads fine

**Comment by hazendaz** [comment]
If you were missing that from tomcat configs, that tells me you are not up to date with tomcat generally as that isn't our requirement but their own.  Could you please take a fresh untouched tomcat and use same jvm and see what you get.  I suspect it will simply work.  Then you could start checking all your setups to see.  

I haven't gotten to test more recent version but last I check the recent versions all have the same settings as I release bundles at my job for secured tomcats and none have lost that setting from core tomcat.

Note that I have released 4.1.5 now.  I'd be happy to fix something here if there is an issue but I'm not seeing any issue.  I'll try bumping up to newer version to double check as soon as I can but may be a couple weeks.  Since this doesn't seem widespread or upvoted, my guess is its some configuration issue with the tomcat instance you have.  That is just a guess though based on missing configs tomcat has long had in place.

**Comment by khravi1979** [comment]
We have built many instances on Tomcat9 after that, but the issue is still the same.
So, why the PROBE URL works after removing probe directory in webapps?

**Comment by hazendaz** [comment]
I cannot answer your specific question.  I can only point back to what I noted previously that you were missing setup and if missing that you are likely missing far more.  Its important when upgrading that tomcat is in sync with tomcat releases.  That isn't something psi probe can help with.  I've since upgraded to latest and no issues but note I also updated all the config changes they have had over the years too.  There are changes from time to time in tomcat configs, those need adhered to on any upgrades.  I haven't run into the issue you are facing otherwise.
